# 2.11.2可恢复的错误 Result

```rust
enum Result<T,E> {
  Ok(T),
  Err(E),
}
```

### unwrap expect


### 传播错误


一个设计良好的系统中，自定义的错误特征，错误之间很可能会存在上下级关系。 std::io::Error 和 std::error:Error

? 自动类型提升

```rust
fn open_file()=>Result<File,Box<dyn std::error::Error>> {
  let mut f = File::open("hello.txt")?;
  Ok(f)
}
```